### PR TITLE
fix(proxy): correct handling of connections

### PR DESCRIPTION
**Pull request by xla**
Long overdue iteration on the Peers core run loop to correctly handle
events from the protocol pertaining to connections. At the same time
reacting appropriately to sync events which have been neglected so far.
There is a range of further improvements that could land subsequently.

### CODE REVIEW

**Review by rudolfs — APPROVED** [review]

File: proxy/coco/src/peer/run_state.rs [hunk 1]
```diff
@@ -394,61 +424,74 @@ impl RunState {
...
             (_, ProtocolEvent::Disconnecting(peer_id)) => {
-                self.connected_peers.remove(&peer_id);
+                if let Some(counter) = self.connected_peers.get_mut(&peer_id) {
+                    *counter -= 1;
+
+                    if *counter == 0 {
+                        self.connected_peers.remove(&peer_id);
+                    }
+                } else {
+                    unreachable!();
```
Comment (FintanH) [inline proxy/coco/src/peer/run_state.rs:493]: Why do we need this `else` branch?

File: proxy/coco/src/peer/run_state.rs [hunk 2]
```diff
@@ -299,7 +301,7 @@ pub struct RunState {
     /// Confiugration to change how input [`Input`]s are interpreted.
     config: Config,
     /// Tracking remote peers that have an active connection.
-    connected_peers: HashSet<PeerId>,
+    connected_peers: HashMap<PeerId, usize>,
```
Comment (FintanH) [inline proxy/coco/src/peer/run_state.rs:304]: I'm not sure I understood the motivation for the `HashMap`. Could you educate me, please? :)

**Review by FintanH — CHANGES_REQUESTED** [review]

File: proxy/coco/src/peer/run_state.rs (hunk 2 above)
Comment (xla) [reply proxy/coco/src/peer/run_state.rs:304]: The peer can receive multiple connections from the same `peer_id`,  a vivid example is the extra connection that is established during syncs (git only exchange). If we just keep track of simply a binary connected status the following order of events can leave the peer in an offline state: `Connect(Peer1) -> Connect(Peer1) -> Disconnect(Peer1)`.

File: proxy/coco/src/peer/run_state.rs (hunk 1 above)
Comment (xla) [reply proxy/coco/src/peer/run_state.rs:493]: This branch should never been hit, hence the unreachable. Haven't found a way to encode this invariance.

File: proxy/coco/src/peer/run_state.rs (hunk 2 above)
Comment (FintanH) [reply proxy/coco/src/peer/run_state.rs:304]: Ahhhh very good :clap: Can you expand on the docs for `connected_peers` to outline this case? Some prose like that would have illuminated my mind immediately :bow:

File: proxy/coco/src/peer/run_state.rs (hunk 1 above)
Comment (FintanH) [reply proxy/coco/src/peer/run_state.rs:493]: But that branch doesn't need to be included. So could we log an error instead or something? Like:

```rust
log::error!("The impossible has happened, somehow we disconnected from '{}' without already being connected to them", peer_id);
```

File: proxy/coco/src/peer/run_state.rs (hunk 2 above)
Comment (xla) [reply proxy/coco/src/peer/run_state.rs:314]: 1c73907

**Review by FintanH — APPROVED** [review]
Nice one :metal:

**Review by rudolfs — APPROVED** [review]

**Pull request merged**